README.md
HMTL5 Player plugin for Meister
=========

This plugin allows playback using the HTML5 player. This is only a player plugin so it cannot play media on it's own, it needs a media plugin to play. For example the plugin [BaseMedia](https://github.com/meisterplayer/media-basemedia) 

Getting started
---------

Load the plugin by adding the ```Html5Player``` configuration object to the Meister initialisation options.

``` JavaScript
var meisterPlayer = new Meister('#player', {
    Html5Player: {}
});

// Continue Meister configuration
Meister.setItem({ ... });
Meister.load();
```

Config options
----------

The following options can be set in the Html5Player: {} config.

### checkInterval *[Number]* (default: 200) ###

The interval checks (in ms) if the player is buffering. The player decides if it's buffering if the player is not paused and the currentTime hasn't moved compared to the last position. Higher numbers result in a longer time before the player decides whether or not the player is buffering.

Example:


``` JavaScript
var meisterPlayer = new Meister('#player', {
    Html5Player: {
        checkInterval: 300,
    }
});
```

### playerSeekDebounce *[Number]* (default: 500) ###

Due to differences in how browsers handle seeking through video it is necessary to debounce the trigger for the 'playerSeekComplete' event. Higher values reduce the number of duplicate events at the cost of losing some accuracy. By default this trigger is debounced for 500 milliseconds.

Example:

``` JavaScript
var meisterPlayer = new Meister('#player', {
    Html5Player: {
        playerSeekDebounce: 700,
    }
});
```



### audioOnly *[Boolean]* (default: false) ###

Set the player in audio only mode. This results in using only the ```<audio>``` tag for playback instead of using the ```<video>``` tag. If you want this option only to set per item you can use the [mediaType]() property on ```setItem();```

Example:

``` JavaScript
var meisterPlayer = new Meister('#player', {
    Html5Player: {
        audioOnly: true,
    }
});
```

### defaultAudioImage *[String]* (default: '') ###

Sets the image for audioOnly streams. This way you can create a visual indicating the stream is audio only.

Example: 

``` JavaScript
var meisterPlayer = new Meister('#player', {
    Html5Player: {
        audioOnly: true,
        defaultAudioImage: 'https://example.com/images/audio.png',
    }
});
```

### iosPlaysInline *[Boolean]* (default: false) ###

Adds the attribute playsinline to the ```<video>``` or ```<audio>``` element. This way you can play the video inline on iOS devices. 

Example: 

``` JavaScript
var meisterPlayer = new Meister('#player', {
    Html5Player: {
        iosPlaysInline: true
    }
});
```

### startMuted *[Boolean]* (default: false) ###

Options to start the player in muted mode.

Example: 

``` JavaScript
var meisterPlayer = new Meister('#player', {
    Html5Player: {
        startMuted: true
    }
});
```

### posterImage *[String]* (default: undefined) ###

A URL indicating a poster frame to show until the user plays or seeks. If this attribute isn't specified, nothing is displayed until the first frame is available; then the first frame is shown as the poster frame.

Example:

``` JavaScript
var meisterPlayer = new Meister('#player', {
    Html5Player: {
        posterImage: 'http://example.com/poster.png'
    }
});
```

### enableKeyBoardShortcuts *[Boolean]* (default: true) ###

Enables the keyboard shorcuts on meister

Example: 

``` JavaScript
var meisterPlayer = new Meister('#player', {
    Html5Player: {
        enableKeyBoardShortcuts: false // Disables the keyboard shorcuts
    }
});
```


Item options
--------

The following options are available for per item configuration.

### mediaType *[String]* (default: undefined) ###

The same as the config option ```audioOnly``` only this can be set per item.

Example: 

``` JavaScript
var meisterPlayer = new Meister('#player', {
    Html5Player: {}
});

meisterPlayer.setItem({
    src: 'INSERT_SRC_HERE',
    mediaType: 'audio',
    type: 'mp4'
});
```

### crossorigin *[String]* (default: undefined) ###

Sets the crossorigin attribute on the ```<video>``` or ```<audio>``` element. For more information on this attribute and what properties can be set, you can visit the [MDN Documentation](https://developer.mozilla.org/en-US/docs/Web/HTML/CORS_settings_attributes)

Example: 

``` JavaScript
var meisterPlayer = new Meister('#player', {
    Html5Player: {}
});

meisterPlayer.setItem({
    src: 'INSERT_SRC_HERE',
    crossorigin: 'anonymous',
    type: 'mp4'
});
```



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="5.9.3"></a>
## [5.9.3](https://github.com/meisterplayer/player-html5player/compare/v5.9.2...v5.9.3) (2018-12-17)


### Bug Fixes

* **events:** No longer fire firstPlay event on ad resume ([ea51e76](https://github.com/meisterplayer/player-html5player/commit/ea51e76))



<a name="5.9.2"></a>
## [5.9.2](https://github.com/meisterplayer/player-html5player/compare/v5.9.1...v5.9.2) (2018-12-13)


### Bug Fixes

* **play:** Remove double play to prevent black screens ([49babc8](https://github.com/meisterplayer/player-html5player/commit/49babc8))



<a name="5.9.1"></a>
## [5.9.1](https://github.com/meisterplayer/player-html5player/compare/v5.9.0...v5.9.1) (2018-11-20)


### Bug Fixes

* **replay:** Fix issue where replay was not triggered correctly ([b4083a6](https://github.com/meisterplayer/player-html5player/commit/b4083a6))



<a name="5.9.0"></a>
# [5.9.0](https://github.com/meisterplayer/player-html5player/compare/v5.8.0...v5.9.0) (2018-10-30)


### Bug Fixes

* **events:** Fix issue where firstPlay was not triggered after new item was loaded ([cbbfed1](https://github.com/meisterplayer/player-html5player/commit/cbbfed1))
* **replay:** Fix issue where replay was triggering on the wrong moment ([8a25bc5](https://github.com/meisterplayer/player-html5player/commit/8a25bc5))


### Features

* **error:** Add HDCP error ([dcdc3d7](https://github.com/meisterplayer/player-html5player/commit/dcdc3d7))



<a name="5.8.0"></a>
# [5.8.0](https://github.com/meisterplayer/player-html5player/compare/v5.7.0...v5.8.0) (2018-06-12)


### Bug Fixes

* **events:** Reduce the number of seekCompleted events on Edge ([1acee93](https://github.com/meisterplayer/player-html5player/commit/1acee93))


### Features

* **events:** Add configurable debounce on the playerSeekComplete event ([59f9f11](https://github.com/meisterplayer/player-html5player/commit/59f9f11))



# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="5.7.0"></a>
# [5.7.0](https://github.com/meisterplayer/player-html5player/compare/v5.6.2...v5.7.0) (2018-05-01)


### Bug Fixes

* **buffering:** Fix issue where playerBufferedEnough was sometimes not called ([3edaca0](https://github.com/meisterplayer/player-html5player/commit/3edaca0))


### Features

* **events:** Add playerSeekComplete as a replacement for seeked ([030092f](https://github.com/meisterplayer/player-html5player/commit/030092f))



<a name="5.6.2"></a>
## [5.6.2](https://github.com/meisterplayer/player-html5player/compare/v5.6.1...v5.6.2) (2018-02-26)


### Bug Fixes

* **replay:** Fix issue where unnecessary buffering was triggered ([b8fcc4a](https://github.com/meisterplayer/player-html5player/commit/b8fcc4a))



<a name="5.6.1"></a>
## [5.6.1](https://github.com/meisterplayer/player-html5player/compare/v5.6.0...v5.6.1) (2018-02-05)


### Bug Fixes

* **events:** moved setting shouldTriggerReplay switch to playerEnd event ([b2eb7f6](https://github.com/meisterplayer/player-html5player/commit/b2eb7f6))



<a name="5.6.0"></a>
# [5.6.0](https://github.com/meisterplayer/player-html5player/compare/v5.5.1...v5.6.0) (2018-02-05)


### Bug Fixes

* **playback:** DisablePauseWithLive should now work properly ([a4a3970](https://github.com/meisterplayer/player-html5player/commit/a4a3970)), closes [meisterplayer/meisterplayer#30](https://github.com/meisterplayer/meisterplayer/issues/30)


### Features

* **audio:** Use <img> tag instead of <div> for background images ([cab99b5](https://github.com/meisterplayer/player-html5player/commit/cab99b5))



<a name="5.5.1"></a>
## [5.5.1](https://github.com/meisterplayer/player-html5player/compare/v5.5.0...v5.5.1) (2017-11-07)


### Bug Fixes

* **disablePauseWithLive:** Fix issue with keyboard controls and disabled pause ([e0bbeb0](https://github.com/meisterplayer/player-html5player/commit/e0bbeb0))



<a name="5.5.0"></a>
# [5.5.0](https://github.com/meisterplayer/player-html5player/compare/v5.4.0...v5.5.0) (2017-08-28)


### Features

* **enableKeyBoardShortcuts:** Add the option to disable the keyboard shortcuts ([8babbdb](https://github.com/meisterplayer/player-html5player/commit/8babbdb))



<a name="5.4.0"></a>
# [5.4.0](https://github.com/meisterplayer/player-html5player/compare/v5.3.2...v5.4.0) (2017-06-27)


### Features

* **pluginVersion:** Add pluginVersion to class ([aa77c17](https://github.com/meisterplayer/player-html5player/commit/aa77c17))



<a name="5.3.2"></a>
## [5.3.2](https://github.com/meisterplayer/player-html5player/compare/v5.3.1...v5.3.2) (2017-06-23)


### Bug Fixes

* **html5player:** replaced append with appendChild, MSIE does not like append ([8a75357](https://github.com/meisterplayer/player-html5player/commit/8a75357))



<a name="5.3.0"></a>
# [5.3.0](https://github.com/meisterplayer/player-html5player/compare/v5.2.0...v5.3.0) (2017-06-20)


### Bug Fixes

* **itemTimeInfo:** Remove call to super method that does not exist ([bc8ba85](https://github.com/meisterplayer/player-html5player/commit/bc8ba85))
* **load:** Fix issue where itemLoaded was triggered while not using Html5Player ([fe2a9b3](https://github.com/meisterplayer/player-html5player/commit/fe2a9b3))


### Features

* **events:** Add playerCanPlay ([796d968](https://github.com/meisterplayer/player-html5player/commit/796d968))



<a name="5.2.0"></a>
# [5.2.0](https://github.com/meisterplayer/player-html5player/compare/v5.1.0...v5.2.0) (2017-05-29)


### Bug Fixes

* **live:** Livestreams can be paused using keyboard controls ([09cb2c9](https://github.com/meisterplayer/player-html5player/commit/09cb2c9)), closes [#30](https://github.com/meisterplayer/player-html5player/issues/30)


### Features


test/Html5Player.test.js
import '@meisterplayer/meister-mock';
import Html5Player from '../src/js/Html5Player';

const PLUGIN_NAME = 'Html5Player';

describe('Html5Player class', () => {
    test(`pluginName should be ${PLUGIN_NAME}`, () => {
        expect(Html5Player.pluginName).toBe(PLUGIN_NAME);
    });

    test('pluginVersion should return a version string', () => {
        // Version should match the SemVer pattern (e.g. 2.11.9)
        expect(Html5Player.pluginVersion).toMatch(/\d+\.\d+\.\d+/);
    });
});


gulpfile.js
const gulp = require('gulp');
const tripleGulp = require('meister-js-dev').gulp;
const webpackTask = require('meister-gulp-webpack-tasks');

// Building tasks.
const MODULE_NAME = 'Html5player';

// building tasks
gulp.task('build', (done) => {
    const bundleConfig = webpackTask.createConfig('./index.js', `build/${MODULE_NAME}.js`, false);
    const bundleCompiler = webpackTask.createCompiler(bundleConfig);

    webpackTask.createBuildTask(bundleCompiler)(done);
});

gulp.task('build:dist', (done) => {
    const bundleConfigDist = webpackTask.createConfig('./index.js', `dist/${MODULE_NAME}.js`, false);
    const bundleCompilerDist = webpackTask.createCompiler(bundleConfigDist);

    webpackTask.createBuildTask(bundleCompilerDist)(done);
});

gulp.task('build:min', (done) => {
    const bundleConfigMin = webpackTask.createConfig('./index.js', `dist/${MODULE_NAME}.min.js`, true);
    const bundleCompilerMin = webpackTask.createCompiler(bundleConfigMin);

    webpackTask.createBuildTask(bundleCompilerMin)(done);
});

// Documentation tasks.
gulp.task('js-docs', tripleGulp.jsdocModule.createGenerateDocs(['./src/**/*.js'], './docs/js-docs'));

// Versioning tasks.
gulp.task('bump-version', tripleGulp.versioningModule.createBumpVersion('./package.json'));

// Changelog tasks.
gulp.task('changelog', tripleGulp.changelogModule.createGenerateLog('./CHANGELOG.md'));

// Cleaning tasks
gulp.task('clean:build', tripleGulp.cleanModule.createClean('./build/*'));
gulp.task('clean:dist', tripleGulp.cleanModule.createClean('./dist/*'));


src/js/utils/Keyboard.js
class Keyboard {

    // function keys
    static get F1() { return 112; }
    static get F2() { return 113; }
    static get F3() { return 114; }
    static get F4() { return 115; }
    static get F5() { return 116; }
    static get F6() { return 117; }
    static get F7() { return 118; }
    static get F8() { return 118; }
    static get F9() { return 119; }
    static get F10() { return 120; }
    static get F11() { return 121; }
    static get F12() { return 122; }

    // Digits
    static get Digit1() { return 49; }
    static get Digit2() { return 50; }
    static get Digit3() { return 51; }
    static get Digit4() { return 52; }
    static get Digit5() { return 53; }
    static get Digit6() { return 54; }
    static get Digit7() { return 55; }
    static get Digit8() { return 56; }
    static get Digit9() { return 57; }
    static get Digit0() { return 48; }

    // special chars
    static get Minus() { return 189; }
    static get Equal() { return 187; }
    static get Comma() { return 188; }
    static get Period() { return 190; }
    static get Slash() { return 191; }
    static get Dash() { return 189; }
    static get Semicolon() { return 186; }
    static get Quote() { return 222; }
    static get BracketLeft() { return 219; }
    static get BracketRight() { return 221; }
    static get Backquote() { return 192; }
    static get Accent() { return 192; }
    static get BackSlash() { return 220; }

    // control keys
    static get Home() { return 36; }
    static get End() { return 35; }
    static get PageUp() { return 33; }
    static get PageDown() { return 34; }
    static get ArrowRight() { return 39; }
    static get ArrowDown() { return 40; }
    static get ArrowLeft() { return 37; }
    static get ArrowUp() { return 38; }
    static get Shift() { return 16; }
    static get Control() { return 17; }
    static get Alt() { return 18; }
    static get Meta() { return 91; }
    static get Space() { return 32; }
    static get Escape() { return 27; }
    static get CapsLock() { return 20; }
    static get Tab() { return 9; }
    static get Fn() { return 119; }
    static get NumLock() { return 144; }
    static get ScrollLock() { return 145; }
    static get Enter() { return 66; }
    static get NumpadEnter() { return 160; }
    static get Backspace() { return 67; }
    static get Clear() { return 28; }
    static get Delete() { return 112; }
    static get Insert() { return 124; }
    static get Help() { return 259; }
    static get Pause() { return 121; }
    static get BrightnessDown() { return 220; }
    static get BrightnessUp() { return 221; }
    static get Eject() { return 129; }
    static get Power() { return 26; }
    static get PrintScreen() { return 120; }
    static get WakeUp() { return 223; }
    static get Standby() { return 224; }

    // Regular Chars
    static get KeyQ() { return 81; }
    static get KeyW() { return 87; }
    static get KeyE() { return 69; }
    static get KeyR() { return 82; }
    static get KeyT() { return 84; }
    static get KeyY() { return 89; }
    static get KeyU() { return 85; }
    static get KeyI() { return 73; }
    static get KeyO() { return 79; }
    static get KeyP() { return 80; }
    static get KeyA() { return 65; }
    static get KeyS() { return 83; }
    static get KeyD() { return 68; }
    static get KeyF() { return 70; }
    static get KeyG() { return 71; }
    static get KeyH() { return 72; }
    static get KeyJ() { return 74; }
    static get KeyK() { return 75; }
    static get KeyL() { return 76; }
    static get KeyZ() { return 90; }
    static get KeyX() { return 88; }
    static get KeyC() { return 67; }
    static get KeyV() { return 86; }
    static get KeyB() { return 66; }
    static get KeyN() { return 78; }
    static get KeyM() { return 77; }

    // media keys
    static get Mute() { return 173; }
    static get VolumeDown() { return 174; }
    static get VolumeUp() { return 175; }
    static get NextTrack() { return 176; }
    static get PreviousTrack() { return 177; }
    static get Stop() { return 182; }
    static get PlayPause() { return 179; }

}

export default Keyboard;


src/js/utils/KeyboardHandler.js
class KeyboardHandler {

    constructor(container, eventHandler) {
        this.container = container;
        this.eventHandler = eventHandler;

        this.container.addEventListener('keyup', this.handleUp.bind(this));
        this.container.addEventListener('keydown', this.handleDown.bind(this));

        this.container.setAttribute('tabindex', '0');

        this.eventHandler.on('playerFullscreen', () => {
            this.container.focus();
        });
    }

    /**
     * Default method onKey, uses on keyUp to actually trigger, has keyDown defaulting to preventDefault
     */
    onKey(key, callBack) {
        const keyArray = Array.isArray(key) ? key : [key];

        keyArray.forEach((someKey) => {
            this.onKeyUp(someKey, callBack);
            this.onKeyDown(someKey, (e) => {
                e.preventDefault();
            });
        });
    }

    onKeyUp(key, callBack) {
        this.eventHandler.on(`up${key}`, callBack);
    }

    onKeyDown(key, callBack) {
        this.eventHandler.on(`down${key}`, callBack);
    }

    handleUp(e) {
        this.eventHandler.trigger(`up${e.keyCode}`, e);
    }

    handleDown(e) {
        this.eventHandler.trigger(`down${e.keyCode}`, e);
    }

}

export default KeyboardHandler;


src/js/Html5Player.js
import Keyboard from './utils/Keyboard';
import KeyboardHandler from './utils/KeyboardHandler';
import createDebounce from './utils/debounce';
import packageJson from '../../package.json';

// Check every 200 ms, lower returns false positives and higher becomes too unresponsive.
const CHECK_INTERVAL = 200;
const PLAYER_SEEK_DEBOUNCE = 500;

class Html5Player extends Meister.PlayerPlugin {
    constructor(config, meister) {
        super(config, meister);

        this.config.enableKeyBoardShortcuts = (typeof this.config.enableKeyBoardShortcuts === 'undefined' ? true : this.config.enableKeyBoardShortcuts);

        this.isLoaded = false;
        this.seekToWhenLoaded = null;

        // Monitoring progress.
        this.bufferingMonitor = null;
        this.CHECK_INTERVAL = this.config.checkInterval || CHECK_INTERVAL;
        this.lastPlayPos = 0;
        this.buffering = false;
        this.playerPlayEvent = null;
        this.playerPauseEvent = null;
        this.hasNoDRM = false;
        this.currentItem = null;

        this.scrubbingInProgress = false;
        const playerSeekedDebounce = Number.isFinite(this.config.playerSeekDebounce) ? this.config.playerSeekDebounce : PLAYER_SEEK_DEBOUNCE;
        this.debounce = createDebounce(playerSeekedDebounce);

        this.meister.on('playlistMetadata', (item) => {
            this.currentItem = item;
        });
    }

    static get pluginName() {
        return 'Html5Player';
    }

    static get pluginVersion() {
        return packageJson.version;
    }

    // eslint-disable-next-line class-methods-use-this
    isTypeSupported(type) {
        if (type === 'html5') {
            return true;
        }

        return false;
    }

    // eslint-disable-next-line class-methods-use-this
    preloadContent() {
        // It seems that the mediaElement.load() is not needed. Replaced with a hard resolve;
        return Promise.resolve();
    }

    /**
     * Load the global settings into the html5 video element
     */
    load(mediaItem) {
        super.load(mediaItem);

        this.mediaElement = document.createElement('video');
        if (this.meister.config.audioOnly || mediaItem.mediaType === 'audio') {
            // TODO create own plugin for audio player
            this.backgroundImage = document.createElement('img');
            this.defaultImageUrl = this.meister.config.defaultAudioImage || '';
            this.backgroundImage.src = this.defaultImageUrl;
            this.backgroundImage.setAttribute('class', 'pf-audio-image');

            this.wrapper.appendChild(this.backgroundImage);
            this.mediaElement = document.createElement('audio');
        }

        // this.mediaElement.id = this.meister.config.playerID + Math.random();
        this.mediaElement.width = this.config.width;
        this.mediaElement.height = this.config.height;
        this.mediaElement.setAttribute('class', this.meister.config.playerClass);

        if (this.config.posterImage) {
            this.mediaElement.setAttribute('poster', this.config.posterImage);
        }

        // for doubleclick enabled behavior on iOS the playsinline attribute must be true
        if (this.meister.config.iosPlaysInline) {
            this.mediaElement.setAttribute('playsinline', true);
        }

        // this.mediaElement.controls = this.meister.config.controls;
        this.mediaElement.muted = this.meister.config.startMuted;

        this.wrapper.appendChild(this.mediaElement);

        this.meister.on('playerPlay', (e) => {
            // Advertisements don't count as actual playback
            if (e && e.type === 'ad') {
                return;
            }

            // Replays are when an end event has been triggered and the user clicks on play again.
            if (this.shouldTriggerReplay) {
                // Make sure that the replay event is only triggered when the src is the same as the current item.
                // Otherwise it wouldn't be a replay.
                if (mediaItem.src !== this.currentItem.src) {
                    return;
                }

                this.meister.trigger('playerReplay', {});
                this.lastPlayPos = 0;
                this.shouldTriggerReplay = false;
                return;
            }

            if (this.firstPlayTriggered) return;
            this.firstPlayTriggered = true;

            this.meister.trigger('playerFirstPlay', {});
        }, 'Html5Player');


        this.mediaElement.addEventListener('play', () => {
            this.meister.trigger('playerPlay', this.playerPlayEvent);
            this.playerPlayEvent = null;
        });

        this.mediaElement.addEventListener('pause', () => {
            this.meister.trigger('playerPause', this.playerPauseEvent);
            this.playerPauseEvent = null;
        });

        this.mediaElement.addEventListener('playing', () => this.meister.trigger('playerPlaying'));

        this.mediaElement.addEventListener('ended', () => {
            this.meister.trigger('playerEnd');
        });

        this.mediaElement.addEventListener('error', (error) => {
            if (this.mediaElement.error.code === this.mediaElement.error.MEDIA_ERR_SRC_NOT_SUPPORTED) {
                if (this.mediaElement.error.msExtendedCode !== undefined && this.mediaElement.error.msExtendedCode === -1071241984) {
                    // This means the browser does not have the correct video card drivers to play the content.
                    // This currently only happens in edge.
                    this.meister.error('Driver does not support HDCP', Meister.ErrorCodes.HDCP_NOT_SUPPORTED || 'MSTR-1002');
                }
            } else if (this.mediaElement.networkState === HTMLMediaElement.NETWORK_NO_SOURCE) {
                this.meister.error('Media not found', Meister.ErrorCodes.NO_MEDIA_FOUND);
            }

            this.meister.trigger('playerError', { mediaError: this.mediaElement.error });
        });

        this.mediaElement.addEventListener('seeked', () => this.meister.trigger('_playerSeek'));
        this.mediaElement.addEventListener('seeking', () => this.meister.trigger('playerSeeking'));
        this.mediaElement.addEventListener('timeupdate', () => this.meister.trigger('_playerTimeUpdate'));
        this.mediaElement.addEventListener('progress', () => this.meister.trigger('playerProgress', this.mediaElement));
        this.mediaElement.addEventListener('loadedmetadata', () => this.meister.trigger('playerLoadedMetadata'));
        this.mediaElement.addEventListener('durationchange', () => this.meister.trigger('playerDurationChange'));
        this.mediaElement.addEventListener('volumechange', () => this.meister.trigger('playerVolumeChange'));
        this.mediaElement.addEventListener('canplay', () => this.meister.trigger('playerCanPlay'));

        this.mediaElement.addEventListener('loadedmetadata', () => {
            this.isLoaded = true;
            if (this.seekToWhenLoaded) {
                this.currentTime = this.seekToWhenLoaded;
                this.seekToWhenLoaded = null;
            }
        });

        /**
         * Creates a listener for when a seek has been completed
         * This was necassary since some browsers (for example IE/Edge)
         * would not have the 'seeked' event correctly implemented
         * This fixes that issue by listening for each event in a chain
         */
        const listenForPlayerSeekComplete = () => {
            this.meister.one('playerSeeking', () => {
                this.meister.one('playerSeek', () => {
                    if (!this.scrubbingInProgress) {
                        this.debounce(() => {
                            this.meister.trigger('playerSeekComplete');
                        });
                    }

                    listenForPlayerSeekComplete();
                });
            });
        };

        listenForPlayerSeekComplete();

        /**
         * Due to scrubbing firing multiple seek events we need to keep track
         * of the scrubbing state.
         */
        this.on('startScrubbing', () => {
            this.scrubbingInProgress = true;
        });

        this.on('endScrubbing', () => {
            // It's possible to stop scrubbing after the last seek completed,
            // to prevent 'losing' this we fire complete here as well.
            this.debounce(() => {
                this.meister.trigger('playerSeekComplete');
            });

            this.scrubbingInProgress = false;
        });

        // keyboard handling
        if (this.config.enableKeyBoardShortcuts) {
            const kb = new KeyboardHandler(this.meister.container, this.meister.eventHandler);
            kb.onKey([Keyboard.Space, Keyboard.Pause, Keyboard.PlayPause, Keyboard.Stop], this.onSpace.bind(this));
            kb.onKey(Keyboard.NextTrack, this.nextTrack.bind(this));
            kb.onKey(Keyboard.PreviousTrack, this.previousTrack.bind(this));

            kb.onKey(Keyboard.ArrowLeft, this.seekBack.bind(this));
            kb.onKey(Keyboard.ArrowRight, this.seekForward.bind(this));
            kb.onKey(Keyboard.KeyF, this.onKeyF.bind(this));
        }

        // Buffering event.
        this.bufferingMonitor = setInterval(this.monitorBuffering.bind(this), this.CHECK_INTERVAL);

        // Reset nudge counter.
        this.on('itemLoaded', this.onItemLoaded.bind(this));
        this.on('itemUnloaded', this.onItemUnloaded.bind(this));

        this.meister.on('itemTimeInfo', (timeInfo) => {
            this.isLive = timeInfo.isLive;
        });

        // this.on('itemTimeInfo', (timeInfo) => { this.onItemTimeInfo(timeInfo); }); // handled in the player object
        this.on('playerEnd', () => {
            // moved from mediaElement.ended because this event could be temporary disabled and this is not checked in mediaElement.ended
            this.shouldTriggerReplay = true;
        });

        this.meister.trigger('playerCreated');
    }

    onItemLoaded(event) {
        if (event.item.crossorigin) {
            this.mediaElement.setAttribute('crossorigin', event.item.crossorigin);
        } else {
            this.mediaElement.removeAttribute('crossorigin');
        }

        if (this.meister.config.audioOnly || event.item.mediaType === 'audio') {
            if (event.item.backgroundImage) {
                this.backgroundImage.style['background-image'] = `url(${event.item.backgroundImage})`;
            } else {
                this.backgroundImage.style['background-image'] = `url(${this.defaultImageUrl})`;
            }
        }
    }

    onItemUnloaded() {
        this.canNudge = 0;
        this.firstPlayTriggered = false;
        this.scrubbingInProgress = false;
    }

    monitorBuffering() {
        const currentPlayPos = this.mediaElement.currentTime;

        // Convert ms to s.
        const offset = 1 / this.CHECK_INTERVAL;

        if (!this.mediaElement.paused) {
            // Special check for special Samsung stock browser which reports currentTime as 0 for icecast streams.
            if (this.meister.browser.isSamsung && this.mediaElement.readyState === this.mediaElement.HAVE_ENOUGH_DATA && currentPlayPos === 0) {
                this.meister.trigger('playerBufferedEnough');
            } else if (!this.buffering && currentPlayPos < this.lastPlayPos + offset) {
                this.buffering = true;
                this.meister.trigger('playerBuffering');
                this.meister.trigger('showLoading', {
                    code: 'VIDEO_BUFFERING',
                });
            } else if (this.buffering && currentPlayPos > this.lastPlayPos + offset) {
                this.buffering = false;
                this.meister.trigger('playerBufferedEnough');
            }
        } else if (this.buffering) {
            this.buffering = false;
            this.meister.trigger('playerBufferedEnough');
        }

        this.lastPlayPos = currentPlayPos;
    }

    unload() {
        clearInterval(this.bufferingMonitor);
        if (this.mediaElement) { this.mediaElement.remove(); }

        this.canNudge = 0;
        this.firstPlayTriggered = false;

        this.meister.trigger('playerDestroyed');
        super.unload();
    }

    set currentSrc(url) {
        if (!this.mediaElement) { return; }

        this.isLoaded = false;
        this.mediaElement.src = url;
    }

    get currentSrc() {
        if (!this.mediaElement) { return null; }

        return this.mediaElement.src;
    }

    get volume() {
        if (!this.mediaElement) { return null; }

        return this.mediaElement.volume;
    }

    set volume(volume) {
        if (!this.mediaElement) { return; }

        this.mediaElement.volume = volume;
    }

    get muted() {
        if (!this.mediaElement) { return null; }

        return this.mediaElement.muted;
    }

    set muted(muted) {
        if (!this.mediaElement) { return; }

        this.mediaElement.muted = muted;
    }

    play(e) {
        this.playerPlayEvent = e; // keep reference to e in class so we can pass it to playerPlay in event stack

        // Player unloaded.
        if (!this.mediaElement) { return; }

        // Chrome 50 introduces promise on media.play()
        this.mediaElement.play();
    }

    pause(e) {
        this.playerPauseEvent = e;

        if (!this.mediaElement) { return; }

        this.mediaElement.pause();
    }

    isPauseDisabled() {
        return !!this.isLive && !!this.meister.config.disablePauseWithLive;
    }

    get duration() {
        if (!this.mediaElement) { return null; }

        return this.mediaElement.duration;
    }

    get playing() {
        if (!this.mediaElement) { return null; }

        return !this.mediaElement.paused;
    }

    get currentTime() {
        if (!this.mediaElement) { return null; }

        return this.mediaElement.currentTime;
    }

    set currentTime(time) {
        if (!this.mediaElement) { return; }

        if (this.isLoaded) {
            this.mediaElement.currentTime = time;
        } else {
            this.seekToWhenLoaded = time;
        }
    }

    onSpace(e) {
        e.preventDefault();
        if (this.isPauseDisabled()) return;
        if (this.meister.playing === true) {
            this.meister.pause();
        } else {
            this.meister.play();
        }
    }

    nextTrack(e) {
        e.preventDefault();
        this.meister.trigger('playlistNext');
    }

    previousTrack(e) {
        e.preventDefault();
        this.meister.trigger('playlistPrevious');
    }


    seekBack(e) {
        e.preventDefault();
        if (this.isPauseDisabled()) return;
        this.meister.trigger('requestSeek', {
            timeOffset: -5,
        });
    }

    seekForward(e) {
        e.preventDefault();
        if (this.isPauseDisabled()) return;
        this.meister.trigger('requestSeek', {
            timeOffset: 5,
        });
    }

    onKeyF(e) {
        e.preventDefault();
        if (this.meister.isFullscreen) {
            this.meister.cancelFullscreen();
        } else {
            this.meister.requestFullscreen();
        }
    }
}

Meister.registerPlugin(Html5Player.pluginName, Html5Player);

export default Html5Player;
